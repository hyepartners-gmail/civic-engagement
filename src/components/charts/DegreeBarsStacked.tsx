"use client";
import { useState, useMemo } from 'react';
import { BarChart, Bar, XAxis, YAxis, CartesianGrid, Tooltip, Legend, ResponsiveContainer } from 'recharts';

interface DegreeBarsStackedProps {
  data: { year: number; hdd: number | null; cdd: number | null }[];
  showHdd: boolean;
  showCdd: boolean;
}

export default function DegreeBarsStacked({ data, showHdd, showCdd }: DegreeBarsStackedProps) {
  // Filter and prepare data based on visibility settings
  const filteredData = useMemo(() => {
    return data.map(item => ({
      year: item.year,
      hdd: showHdd ? item.hdd : null,
      cdd: showCdd ? item.cdd : null,
    })).filter(item => 
      (showHdd && item.hdd !== null) || 
      (showCdd && item.cdd !== null)
    );
  }, [data, showHdd, showCdd]);

  // Calculate domain dynamically based on visible data
  const yDomain = useMemo(() => {
    if (filteredData.length === 0) return [0, 100];
    
    const values = filteredData.flatMap(item => [
      item.hdd ?? 0,
      item.cdd ?? 0
    ]).filter(val => val !== null) as number[];
    
    if (values.length === 0) return [0, 100];
    
    const max = Math.max(...values);
    return [0, max * 1.1]; // Add 10% padding
  }, [filteredData]);

  return (
    <div className="w-full h-80">
      <ResponsiveContainer width="100%" height="100%">
        <BarChart
          data={filteredData}
          margin={{ top: 20, right: 30, left: 20, bottom: 40 }}
        >
          <CartesianGrid strokeDasharray="3 3" />
          <XAxis 
            dataKey="year" 
            angle={-45} 
            textAnchor="end" 
            height={60}
            tick={{ fontSize: 12 }}
          />
          <YAxis 
            domain={yDomain}
            tick={{ fontSize: 12 }}
          />
          <Tooltip 
            formatter={(value) => [Number(value).toFixed(1), 'Degree Days']}
            labelFormatter={(year) => `Year: ${year}`}
          />
          <Legend />
          {showHdd && (
            <Bar 
              dataKey="hdd" 
              name="Heat On Cold Days (HDD)" 
              fill="#ef4444" 
              stackId="a" 
            />
          )}
          {showCdd && (
            <Bar 
              dataKey="cdd" 
              name="Cooling on Hot Days (CDD)" 
              fill="#3b82f6" 
              stackId="a" 
            />
          )}
        </BarChart>
      </ResponsiveContainer>
    </div>
  );
}