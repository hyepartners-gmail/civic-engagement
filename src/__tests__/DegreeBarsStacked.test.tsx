import React from 'react';
import { render } from '@testing-library/react';
import DegreeBarsStacked from '@/components/charts/DegreeBarsStacked';

// Mock the recharts components
jest.mock('recharts', () => ({
  BarChart: ({ children }: any) => <div data-testid="bar-chart">{children}</div>,
  Bar: ({ dataKey, name, fill }: any) => <div data-testid={`bar-${dataKey}`} data-name={name} data-fill={fill} />,
  XAxis: () => <div data-testid="x-axis" />,
  YAxis: () => <div data-testid="y-axis" />,
  CartesianGrid: () => <div data-testid="cartesian-grid" />,
  Tooltip: () => <div data-testid="tooltip" />,
  Legend: () => <div data-testid="legend" />,
  ResponsiveContainer: ({ children }: any) => <div data-testid="responsive-container">{children}</div>,
}));

describe('DegreeBarsStacked', () => {
  const mockData = [
    { year: 2000, hdd: 1000, cdd: 2000 },
    { year: 2001, hdd: 950, cdd: 2100 },
    { year: 2002, hdd: 900, cdd: 2200 },
  ];

  it('renders stacked bars with both HDD and CDD visible', () => {
    const { getByTestId } = render(
      <DegreeBarsStacked data={mockData} showHdd={true} showCdd={true} />
    );

    expect(getByTestId('bar-chart')).toBeInTheDocument();
    expect(getByTestId('bar-hdd')).toBeInTheDocument();
    expect(getByTestId('bar-cdd')).toBeInTheDocument();
    expect(getByTestId('x-axis')).toBeInTheDocument();
    expect(getByTestId('y-axis')).toBeInTheDocument();
    expect(getByTestId('cartesian-grid')).toBeInTheDocument();
    expect(getByTestId('tooltip')).toBeInTheDocument();
    expect(getByTestId('legend')).toBeInTheDocument();
  });

  it('renders only HDD when CDD is hidden', () => {
    const { getByTestId, queryByTestId } = render(
      <DegreeBarsStacked data={mockData} showHdd={true} showCdd={false} />
    );

    expect(getByTestId('bar-hdd')).toBeInTheDocument();
    expect(queryByTestId('bar-cdd')).not.toBeInTheDocument();
  });

  it('renders only CDD when HDD is hidden', () => {
    const { getByTestId, queryByTestId } = render(
      <DegreeBarsStacked data={mockData} showHdd={false} showCdd={true} />
    );

    expect(getByTestId('bar-cdd')).toBeInTheDocument();
    expect(queryByTestId('bar-hdd')).not.toBeInTheDocument();
  });

  it('renders responsive container', () => {
    const { getByTestId } = render(
      <DegreeBarsStacked data={mockData} showHdd={true} showCdd={true} />
    );

    expect(getByTestId('responsive-container')).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { container } = render(
      <DegreeBarsStacked data={mockData} showHdd={true} showCdd={true} />
    );

    expect(container).toMatchSnapshot();
  });
});