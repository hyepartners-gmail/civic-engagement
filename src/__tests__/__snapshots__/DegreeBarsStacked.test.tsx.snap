// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DegreeBarsStacked matches snapshot 1`] = `
<div>
  <div
    class="w-full h-80"
  >
    <div
      data-testid="responsive-container"
    >
      <div
        data-testid="bar-chart"
      >
        <div
          data-testid="cartesian-grid"
        />
        <div
          data-testid="x-axis"
        />
        <div
          data-testid="y-axis"
        />
        <div
          data-testid="tooltip"
        />
        <div
          data-testid="legend"
        />
        <div
          data-fill="#3b82f6"
          data-name="Heating Degree Days (HDD)"
          data-testid="bar-hdd"
        />
        <div
          data-fill="#ef4444"
          data-name="Cooling Degree Days (CDD)"
          data-testid="bar-cdd"
        />
      </div>
    </div>
  </div>
</div>
`;
